### DISCUSSION TITLE
求大佬说一下远程控制alas自行搭建服务器的大概思路

### DISCUSSION THREAD

**Discussion by asdfhdasfhasd** [LmeSzinc/AzurLaneAutoScript#1561] [Q&A] [answered]
根据指示使用frp已经实现了服务器访问内网计算机，同时也使用密钥实现了客户端无密码登录服务器，但是自行搭建那里描述实在是太简略了，小白完全没搞清alas的远程控制是如何实现的。现在使用服务器host:22作为alas配置中的SSHServer，但是提示服务器未连接，不知道是哪里出现了问题。

**Comment by 18870** [comment]
远程访问本质上是 ssh -R 命令转发本地端口到服务器上

https://github.com/LmeSzinc/AzurLaneAutoScript/blob/378ee5e8ab531f941e3e9a5164c0c2f8653ff290/module/webui/remote_access.py#L62-L63

实际上直接连接22端口，转发到服务器上的一个端口也是可以的。但是设计时并没有考虑这个。
frp 是实现转发端口的一种方式，但是在alas里并没有直接的支持，缺点是frp你要自己管理frpc进程。
还是那句话，不建议自己搭建localshare，特别是只有你自己用。localshare设计上是公共性服务，便于开发者获得一个临时公网链接让其他人访问。不如想一下~来pr~怎么把alas内现有的代码改成支持密钥对访问，转发到服务器上特定端口。

***
docker 直接搭建 localshare 参考 https://github.com/wang0618/localshare/blob/master/build.md

由于未知bug，要在alas上直接使用需要把这行
https://github.com/wang0618/localshare/blob/bdf2e1fc9915b6889a1a23eff0a52e5903c02681/main.py#L61
改成 `await process.wait_closed()`

要怎么在docker上改我不清楚，~反正我不用docker~, gl

并且这个 docker 版本会随机生成用户名无关的访问链接
https://github.com/wang0618/localshare/blob/bdf2e1fc9915b6889a1a23eff0a52e5903c02681/main.py#L88

或者不用 docker
参考 https://github.com/wang0618/localshare/tree/master/docker 内文件
你需要在服务器上安装 nginx，参考nginx配置文件模板配置。
运行 localshare，alas内填入localshare的ip端口，浏览器访问nginx的端口。

另外，localshare使用unix转发，意味着你不能在windows服务器上搭建。

**Reply by asdfhdasfhasd** [reply]
感谢回复

**Comment by yzbyzz** [comment, marked as answer]
不知道你的使用场景是怎样的，我说一下我的使用情况。

设备和设置如下：
- 运行模拟器 和 ALAS 的内网计算机
    - 运行模拟器和 ALAS
    - 运行 frpc （frp 设置 http 方式）
- 具有固定公网 ip 的云服务器
    - 运行 frps
    - 运行 Nginx
    - 配置域名 `_YOUR_DOMAIN_` 指向此云服务器，Nginx 将请求转发到 frps 端口 `_FRPS_SERVER_HTTP_PORT_`


参考配置
```
# frpc.ini

[common]
server_addr = _FRPS_SERVER_IP_
server_port = _FRPS_SERVER_PORT_

[web]
type = http
local_ip = 127.0.0.1
local_port = 22267
custom_domains = _YOUR_DOMAIN_
```

```
# frps.ini

[common]
bind_port = _FRPS_SERVER_PORT_
vhost_http_port = _FRPS_SERVER_HTTP_PORT_

# console or real logFile path like ./frps.log
log_file = /var/log/frps.log

# trace, debug, info, warn, error
log_level = info

log_max_days = 3
```

在其他任意电脑或者手机上访问 `_YOUR_DOMAIN_` 即可访问内网计算机上的 ALAS。（和在本地打开 http://127.0.0.1:22267/ 的页面相同。）

**Reply by asdfhdasfhasd** [reply]
感谢回复，我的应用场景就是这样

**Reply by asdfhdasfhasd** [reply]
非常感谢，已经成功了，原来使用frp就不需要开启alas中的 _RemoteAccess_
